optuna_dashboard/_app.py

```diff
@@ -411,6 +412,45 @@ def save_study_note(study_id: int) -> BottleViewReturn:
         response.status = 204  # No content
         return {}
 
+    @app.post("/api/studies/<trial_id:int>/tell")
+    @json_api_view
+    def tell_trial(trial_id: int) -> BottleViewReturn:
+        s = request.json.get("state", None)
+        vs = request.json.get("values", None)
+
+        try:
+            values = [float(v) for v in vs] if vs is not None else vs
+        except ValueError:
+            response.status = 400  # Bad request
+            return {"reason": "You need to pass float castable values"}
+
+        str_to_state = {
+            "Complete": TrialState.COMPLETE,
+            "Pruned": TrialState.PRUNED,
+            "Fail": TrialState.FAIL,
+        }
+        if s not in str_to_state:
+            response.status = 400  # Bad request
+            return {
+                "reason": f"You passed {s} as a state, but only 'Complete', 'Pruned' and 'Fail'"
+                "are acceptable states."
+            }
+        if s == "Complete" and values is None:
+            response.status = 400  # Bad request
+            return {
+                "reason": "When you passed 'Complete' as a state, you also need to specify values."
+            }
+        state = str_to_state[s]
```

Comment:
How about replacing these logic like the following?

https://github.com/optuna/optuna-dashboard/commit/dc0fdb6bddf22732402e4b48d777f9ab295077a5

In the above commit, I made following improvements.

* When given `{'state': 'Complete', 'values': 0}`, this view function raise `TypeError` exception. So we must handle it.
* I would say we can convert str to state and then check `state.is_finished()` to validate the state.

Reply:
Thank you 🙇  I cherry-picked it.